packages/marshal/src/types.js

```diff
@@ -83,28 +83,36 @@
  */
 
 /**
- * @typedef Encoding
+ * @template T
+ * @typedef {{ '@qclass': T }} EncodingClass
+ */
+
+/**
+ * @typedef {EncodingClass<'NaN'> |
+ * EncodingClass<'undefined'> |
+ * EncodingClass<'Infinity'> |
+ * EncodingClass<'-Infinity'> |
+ * EncodingClass<'bigint'> & { digits: string } |
+ * EncodingClass<'@@asyncIterator'> |
+ * EncodingClass<'ibid'> & { index: number } |
+ * EncodingClass<'error'> & { name: string, message: string, errorId?: string } |
+ * EncodingClass<'slot'> & { index: number, iface?: InterfaceSpec } |
+ * EncodingClass<'hilbert'> & { original: Encoding, rest?: Encoding }} EncodingUnion
+ * @typedef {{ [index: string]: Encoding, '@qclass'?: undefined }} EncodingRecord
+ * We exclude '@qclass' as a property in encoding records.
+ * @typedef {EncodingUnion | null | string | boolean | number | EncodingRecord} EncodingElement
+ */
+
+/**
+ * @template T
+ * @typedef {Array<T | NestedArray<T>>} NestedArray Helper type to produce
+ * recursive arrays
+ */
+
+/**
+ * @typedef {EncodingElement | NestedArray<EncodingElement>} Encoding
```

Comment:
The following seems much more intuitive to me. I just tried it and it seems to work.

```suggestion
 * @typedef {Encoding[]} EncodingArray
 * We exclude '@qclass' as a property in encoding records.
 * @typedef {EncodingUnion | null | string | boolean | number | EncodingRecord | EncodingArray} Encoding
```

Reply:
If you hover over `Encoding`, your more straightforward approach results in a type of `any`.

Reply:
Turns out my existing "solution" also did that.  I need to invoke `.d.ts` magic.